### PR TITLE
create learning rate for multi program

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix https://github.com/PaddlePaddle/Paddle/issues/8529

## reason
In fc_gan demo, there are two programs that need to be optimized, but currently, the optimizer will only create learning_rate_var once. So the second program will not find the learning rate var.

## detail
`optimizer.minimize(loss, ...)` will create optimize ops and related variables in the block of `loss`. The current optimizer has a filed `optimizer._global_learning_rate` which will be created only once, in the fc_gan demo, the optimizer.minimize is called twice for two losses in two program, so the second optimizer.minimize cannot find learning rate variable in its Program.

## solution
create learning rate var for each program it optimized.

### CODE REVIEW

File: python/paddle/v2/fluid/optimizer.py [hunk 1]
```diff
@@ -36,10 +36,18 @@ class Optimizer(object):
     """
 
     def __init__(self, learning_rate, global_step=None, regularization=None):
-        assert learning_rate is not None
+        if not isinstance(learning_rate, float) and \
+                not isinstance(learning_rate, framework.Variable):
+            raise ValueError("learning rate should be float or Variable")
```
Comment (reyoung) [inline python/paddle/v2/fluid/optimizer.py]: TypeError

File: python/paddle/v2/fluid/optimizer.py [hunk 2]
```diff
@@ -36,10 +36,18 @@ class Optimizer(object):
...
         self._global_step = global_step
         self.regularization = regularization
-        self._global_learning_rate = learning_rate
+        self._learning_rate = learning_rate
+        # each program should have a independent learning rate
+        # program -> Variable(learning_rate)
+        self._learning_rate_map = defaultdict(lambda: None)
```
Comment (reyoung) [inline python/paddle/v2/fluid/optimizer.py]: There is an API called `get` in `dict`.
default

We can use that API rather than defaultdict.

File: python/paddle/v2/fluid/optimizer.py [hunk 3]
```diff
@@ -48,26 +56,33 @@ def __init__(self, learning_rate, global_step=None, regularization=None):
...
-                             "actual type is %s",
-                             type(self._global_learning_rate))
-
-    @property
-    def global_learning_rate(self):
+        lr = self.global_learning_rate()
+
+        if isinstance(lr, framework.Variable):
+            return
+        else:
```
Comment (reyoung) [inline python/paddle/v2/fluid/optimizer.py]: `else` is not needed, since it has been returned in if block.

Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: `else` is used to do type check and give some useful information for code reader

File: python/paddle/v2/fluid/optimizer.py (hunk 2 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: done

File: python/paddle/v2/fluid/optimizer.py (hunk 1 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: done
